### PR TITLE
Print a list of documents in PDF

### PR DESCRIPTION
**Pull request by panglesd**
Hello,

Here is a proposition to add an option to print a pdf containing the list of routes of a research.
For instance, one can find the list of routes located in an area with a suitable difficulty level, then click on the new `Print` button, which will generate a pdf to be saved on a phone or printed.

To do this, I needed a few changes on the organization for document views. Indeed, I needed to display, in a single view, multiple documents, so each document needed to be a component and not a view. I therefore separated the "document api loading" from the displaying of the document. Here is a list of the changes:
- Each AreaView, WaypointView, ... are now viewed as components. `document-view-mixin.js` reflects this. They now take a `promise`, `documentId`, `documentType`, `fields` and `expectedLang` as props instead of finding/loading it themselves
- There is a new view, `DocumentView`, in charge of choosing the right component (AreaView, WaypointView, ...) to load with the right props (`promise`, `documentId`, ...). The route mechanism for document is updated.
- There is a new view, `PrintDocumentList`, which loads all documents, and when everything is resolved, call `window.print()`. The routes for this new view is `/print/{article, waypoint, ...}`
- A button is added to go to the PrintDocumentList from a search, with the same filter.

I hope that this is the right solution to implement this feature! Let me know what you think, I can adapt it to the maintainer's taste. And thanks for this amazing open source project :)

(There are many other natural improvements, such as adding a table of content, but I leave them for another PR.)

### CODE REVIEW

**Comment by cbeauchesne** [comment]
Hi @panglesd ,

Thank you very much for your PR. I'll do a review asap, ETA Wednesday evening.

File: src/views/documents/DocumentsPrintingView.vue [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+      <div v-for="(doc, index) in promises" :key="index" class="column">
+        <component
+          :is="documentType + '-view'"
+          :document-type="doc.documentType"
+          :version="doc.version"
+          :promise="doc.promise"
+          :document-id="doc.documentId"
+          :expected-lang="doc.expectedLang"
+          :fields="doc.fields"
+        />
```
Comment (cbeauchesne) [inline src/views/documents/DocumentsPrintingView.vue:15]: Page break in a middle of a topo is painfull, we can use this: 

Add a div : `<div class="page-break" />`, and CSS : 

```css
<style scoped>
.page-break {
  page-break-after: always;
}
</style>
```

File: src/views/documents/DocumentsPrintingView.vue [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            ),
+            documentId: doc.document_id,
+            expectedLang: this.$route.params.lang ?? this.$language.current,
+            documentType: this.documentType,
+            version: null,
+            fields: constants.objectDefinitions[this.documentType].fields,
+          };
+        });
+        Promise.all(this.promises.map((doc) => doc.promise)).then(() => {
+          Vue.nextTick(function () {
```
Comment (cbeauchesne) [inline src/views/documents/DocumentsPrintingView.vue:90]:
```suggestion
          this.$nextTick(function () {
```

**Review by cbeauchesne — CHANGES_REQUESTED** [review]
Well, there is room for improvment, especially on the loading phase (hard to know when it's ready, when the API is slow), but that is definitly a very good start!

Thanks you so much for your work.

I made some minor remarks and sugesstions, please let me know if you have any comment on them.

File: src/views/documents/utils/PrintButton.vue [hunk 3]
```diff
@@ -0,0 +1,15 @@
+<template>
+  <router-link :to="path" tag="button" class="button is-primary is-size-7-mobile" @click="exportCsv">
```
[the remark this answers is not in the document]
Comment (panglesd) [reply src/views/documents/utils/PrintButton.vue:2]: It is an unforgivable remnant of a copy-paste. I will remove it!

File: src/views/documents/utils/QueryItems.vue [hunk 4]
```diff
@@ -51,6 +51,7 @@
       />
       <load-user-preferences-button class="is-hidden-tablet query-item-component" />
       <export-csv-button v-if="listMode" class="is-small-mobile"></export-csv-button>
+      <print-button />
```
[the remark this answers is not in the document]
Comment (panglesd) [reply src/views/documents/utils/QueryItems.vue:54]: I agree with hiding the button on mobile, and I will try to improve the display.

The current behaviour is that all documents displayed in the page are included in the print. By default, 30 documents are displayed per page, but the user can decide to display 50 or 100 documents per page, and in this case they will all be included in the print. The user can also print "page number 2" of a search. Maybe there can be a tooltip "Print the documents of this page" on the button?

File: src/views/documents/DocumentsPrintingView.vue (hunk 1 above)
Comment (panglesd) [reply src/views/documents/DocumentsPrintingView.vue:15]: I had in mind that it was an improvement to be able to print your topos without wasting every time half a page... but if you think it is better to add a page break I'll include your suggestion.

The best (for the future?) would be that pressing the print button open a piece of UI where user can choose different printing options (keeping image or not, page break at the end of a topo...)

**Comment by panglesd** [comment]
Thanks a lot @cbeauchesne for your review!

Yes quite a few improvements needed, but I wanted to check the base is OK before working on the details. I will try to address your comments (including the loading phase) as soon as possible, but this might happen only early next week...

File: src/views/documents/utils/QueryItems.vue (hunk 4 above)
[the remark this answers is not in the document]
Comment (cbeauchesne) [reply src/views/documents/utils/QueryItems.vue:54]: Good ergonomic question, I'll ask on forum

File: src/views/documents/DocumentsPrintingView.vue (hunk 1 above)
Comment (cbeauchesne) [reply src/views/documents/DocumentsPrintingView.vue:15]: > The best (for the future?) would be that pressing the print button open a piece of UI where user can choose different printing options (keeping image or not, page break at the end of a topo...)

Indeed. if you have the motivation, it can be the very first option. Actually, I feel that the auto-open of the print page when documents are loaded is a little bit weird. it could be a good way to go: 

* A `div` at the top of the page, with an parameters' set design. It includes : 
  * a tick box with a label `start each document on top of each page`
  * a disabled button with `loading...`
  * when all documents are loaded, the  `loading...` becomes `print`

WDYT ?

**Comment by cbeauchesne** [comment]
> Yes quite a few improvements needed, but I wanted to check the base is OK before working on the details. I will try to address your comments (including the loading phase) as soon as possible, but this might happen only early next week...

The base is perfect ! thank you very much for your help

File: src/views/documents/utils/QueryItems.vue (hunk 4 above)
[the remark this answers is not in the document]
Comment (cbeauchesne) [reply src/views/documents/utils/QueryItems.vue:54]: https://forum.camptocamp.org/t/impression-dune-liste-de-documents-question-dergonomie/288077 

If you do not speak french, i'll translate key points.

**Comment by panglesd** [comment]
Unfortunately there does not seem to be an easy way to generate the page numbers in the table of context in pure css...

Do you think I should rename the table of content to "summary", or delete it right away? I think a summary of all the printed routes is still useful, even without the page numbers.

**Comment by panglesd** [comment]
I am considering looking into how translation work in c2c_ui to have a  french version of this PR's feature.
But before that, @cbeauchesne do you see other modifications to be done?

**Comment by panglesd** [comment]
I'm willing to resume working on this PR, but I might have time for that only in approximately one month.

**Comment by loicperrin** [comment]
Hello @panglesd, have you planned to finalize this PR soon? It is very expected by users :-)

**Comment by panglesd** [comment]
Hello @loicperrin!

Seeing that it seems indeed expected by users, and since I had recently a new idea to make it better, I am indeed motivated to resume working on it!

I have not a lot of time though, but I hope to have the PR in a good state in a few week maximum :)

**Comment by panglesd** [comment]
I will undraft this PR when it's ready for another pass of review.

**Comment by panglesd** [comment]
I think that the PR is now in good shape for reviewing.

I removed the first commit (a refactoring to separate displaying documents from fetching the document) to make the code less intrusive.

I rebased and split the PR in more logical commits.

**Comment by brunobesson** [comment]
Thanks a lot, I'll have a look ASAP!

File: src/js/vue-plugins/router.js [hunk 5]
```diff
@@ -83,6 +84,12 @@ const addDocumentTypeView = function (def, viewComponent, editionComponent) {
     component: DocumentsView,
   });
 
+  routes.push({
+    path: '/print/' + def.documentType + 's',
```
Comment (brunobesson) [inline src/js/vue-plugins/router.js:88]: I wonder if reversing the path wouldn't be more consistent?
```suggestion
    path: def.documentType + 's/print',
```

File: src/views/document/utils/document-view-mixin.js [hunk 6]
```diff
@@ -95,9 +95,11 @@ export default {
      * properties that are deducted from URL
      */
     documentId() {
+      if (this.isDraftView || this.isPrintingView) return this.draft.document_id;
```
Comment (brunobesson) [inline src/views/document/utils/document-view-mixin.js:98]:
```suggestion
      if (this.isDraftView || this.isPrintingView) {
        return this.draft.document_id;
      }
```

We try to avoid such shortcut notations in the code, for readablitity.

File: src/views/document/utils/document-view-mixin.js [hunk 7]
```diff
@@ -95,9 +95,11 @@ export default {
...
       return parseInt(this.$route.params.id, 10);
     },
     documentType() {
+      if (this.isPrintingView) return this.$route.name.split('s-')[0];
```
Comment (brunobesson) [inline src/views/document/utils/document-view-mixin.js:102]:
```suggestion
      if (this.isPrintingView) {
        return this.$route.name.split('s-')[0];
      }
```

File: src/views/documents/utils/QueryItems.vue (hunk 4 above)
[the remark this answers is not in the document]
Comment (brunobesson) [reply src/views/documents/utils/QueryItems.vue:71]: This issue is still not fixed, am I wrong?

File: src/views/documents/utils/PrintingSummary.vue [hunk 8]
```diff
@@ -0,0 +1,48 @@
+<template>
+  <div v-if="documents" class="box">
+    <h1 class="title is-1">Summary</h1>
+    <div v-for="(document, index) in documents.documents" :key="index">
+      <span @click="sendSummaryClick(index)">
```
Comment (brunobesson) [inline src/views/documents/utils/PrintingSummary.vue:5]: There is no indication (e.g. cursor) that clicking on a list item will scroll you there. Also, I suggest to us smooth scrolling to help the user undestand he is scrolling whithin the page, othewise, you might think it opened the document itself. See other review comment.
Also, maybe one should have a shortcut to go back to the top of the page?

File: src/views/documents/utils/PrintingSummary.vue [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+      default: null,
+    },
+    documentType: {
+      type: String,
+      default: 'route',
+    },
+  },
+
+  methods: {
+
```
Comment (brunobesson) [inline src/views/documents/utils/PrintingSummary.vue:35]: useless empy line

File: src/views/documents/utils/PrintingSummary.vue [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    sendSummaryClick(index) {
+      this.$emit("summary-click", index);
+    }
+
```
Comment (brunobesson) [inline src/views/documents/utils/PrintingSummary.vue:39]: useless empty line

File: src/views/documents/DocumentsPrintingView.vue [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    },
+
+    scrollTo(name) {
+      const el = document.querySelector("#toc-entry-"+name);
+      if (el) {
+        const docEl = document.documentElement;
+        const docRect = docEl.getBoundingClientRect();
+        const elRect = el.getBoundingClientRect();
+        const y = elRect.top - docRect.top;
+        window.scrollTo(0, y - 50); // navbar height ...
```
Comment (brunobesson) [inline src/views/documents/DocumentsPrintingView.vue:123]:
```suggestion
        window.scrollTo({ top: y - 50, behavior: 'smooth' }); // navbar height ...
```

File: src/views/documents/utils/PrintButton.vue [hunk 12]
```diff
@@ -0,0 +1,15 @@
+<template>
+  <a target="_blank" :href="path">
+    <button class="button is-primary is-size-7-mobile">Print search results</button>
```
Comment (brunobesson) [inline src/views/documents/utils/PrintButton.vue:3]: image

Maybe we could replace the text with a print icon on mobile, to save some space?

File: src/views/documents/utils/PrintingOptions.vue [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+  },
+
+  computed: {
+    offset() {
+      return parseInt(this.$route.query.offset ?? 0);
+    },
+    total() {
+      return this.promises.length;
+    },
+    allLoaded() {
```
Comment (brunobesson) [inline src/views/documents/utils/PrintingOptions.vue:59]: the name is confusing. shouldn't it be `is-loading` or something similar instead?

File: src/views/documents/DocumentsPrintingView.vue [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+      @set-page-breaks="pageBreaks = $event"
+      @set-image-printing="includeImages = $event"
+      @set-summary-visibility="summaryVisibility = $event"
+    />
+    <printing-summary
+      v-if="summaryVisibility"
+      :documents="documents"
+      :document-type="documentType"
+      @summary-click="scrollTo($event)"
+    />
```
Comment (brunobesson) [inline src/views/documents/DocumentsPrintingView.vue:16]: I wonder if there could be a way to use columns for displaying pagination, options and print button and use less vertical space?
But it could be for another PR.

**Review by brunobesson — COMMENTED** [review]
Thanks for the work @panglesd it works well!

Some general remarks otherwise: 

* there are still UI issues with the button positioning: 
image
* there seems to be a problem with included images which are sometimes cropped 
image
* it's a bit confusing to have an option `Include images` which has no effect on images included within the description. Maybe t could be handled too?

**Comment by panglesd** [comment]
Thanks a lot for the review!

I'll try to address the comments (I might not have time before this weekend), and I'll tell you when it's done.

**Comment by brunobesson** [comment]
You're welcome. Thanks for the work 👍

**Comment by panglesd** [comment]
I think I addressed most of your comments!

There are two that are still "pending":
- cropped image. I tried with the css rule `break-inside: avoid` but it does not seem to work well... Maybe I did something wrong, I can try again later.
- Handling images in description:
> it's a bit confusing to have an option `Include images` which has no effect on images included within the description. Maybe t could be handled too?

I am not sure if we should have an option for this, as images included inside a description are generally more of interest than images in the image section. If we handle this case, we should at least provide the option of including only those in the description.
I currently renamed the option to `Include image section` to better reflect what it does. If you think we should, I can include an additional option `Include image in description`.

Thanks!

File: src/views/documents/utils/PrintingOptions.vue [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+      <div class="column printing-option-section">
+        <h2 class="title is-2">Number of documents</h2>
+        <page-selector :documents="documents" page-type="print" />.
+      </div>
+      <div class="column printing-option-section">
+        <h2 class="title is-2">Other options</h2>
+        <ul>
+          <li><input class="print-option-checkbox" v-model="pageBreaks" type="checkbox" />Add page breaks before each document</li>
+          <li><input class="print-option-checkbox" v-model="imagePrinting" type="checkbox" />Include image section</li>
+          <li><input class="print-option-checkbox" v-model="summaryVisibility" checked type="checkbox" />Include a summary</li>
```
Comment (brunobesson) [inline src/views/documents/utils/PrintingOptions.vue:13]:
```suggestion
          <li><label><input class="print-option-checkbox" v-model="pageBreaks" type="checkbox" />Add page breaks before each document</label></li>
          <li><label><input class="print-option-checkbox" v-model="imagePrinting" type="checkbox" />Include image section</label></li>
          <li><label><input class="print-option-checkbox" v-model="summaryVisibility" checked type="checkbox" />Include a summary</label></li>
```

**Comment by brunobesson** [comment]
> cropped image

I spent 2 minutes on it and couldn't find an obvious solution. Le'ts say we can fix this later.

> I am not sure if we should have an option for this, as images included inside a description are generally more of interest than images in the image section.

I was wondering if we should remove them, but it's not that important. Let's keep it as is for now and see users's feedback later on, if any.

I made one last comment so that checkboxes can be used more easily. Other than that, it's fine for me. I'll merge once you tell me it is ok on your side.

**Comment by panglesd** [comment]
While making another testing pass, I noticed that maps would modify the bbox query argument, updating the search.

So for instance searching, `outings/print?qa=draft,great&bbox=-313650,6588359,202523,6873190` was instantly rewritten in `outings/print?qa=draft,great&bbox=-193317,7142602,-152008,7167790`, retriggering the fetch of all routes and possibly modifying the result. Moreover, moving or zooming on a map was also retriggering the search and route fetch!

I fixed this in the last commit. Once this commit is reviewed, it is good to go on my side!

**Comment by cbeauchesne** [comment]
Thanks a lot @panglesd , that's so great !!
